src/geometry/mixed_lines_set.cpp

```diff
@@ -0,0 +1,167 @@
+
+#include <numeric>
+
+#include "geometry/open_polyline.h"
+#include "geometry/polygon.h"
+#include "geometry/shape.h"
+
+
+namespace cura
+{
+
+Shape MixedLinesSet::offset(coord_t distance, ClipperLib::JoinType join_type, double miter_limit) const
+{
+    if (distance == 0)
+    {
+        // Return a shape that contains only actual polygons
+        Shape result;
+
+        for (const PolylinePtr& line : (*this))
+        {
+            if (const std::shared_ptr<const Polygon> polygon = dynamic_pointer_cast<const Polygon>(line))
+            {
+                result.push_back(*polygon);
+            }
+        }
+
+        return result;
+    }
+    else
+    {
+        Shape polygons;
+        ClipperLib::ClipperOffset clipper(miter_limit, 10.0);
+
+        for (const PolylinePtr& line : (*this))
+        {
+            if (const std::shared_ptr<const Polygon> polygon = dynamic_pointer_cast<const Polygon>(line))
+            {
+                // Union all polygons first and add them later
+                polygons.push_back(*polygon);
+            }
+            else
+            {
+                ClipperLib::EndType end_type;
+
+                if (line->addClosingSegment())
+                {
+                    end_type = ClipperLib::etClosedLine;
+                }
+                else if (join_type == ClipperLib::jtMiter)
+                {
+                    end_type = ClipperLib::etOpenSquare;
+                }
+                else
+                {
+                    end_type = ClipperLib::etOpenRound;
+                }
+
+                clipper.AddPath(line->getPoints(), join_type, end_type);
+            }
+        }
+
+        if (! polygons.empty())
+        {
+            polygons = polygons.unionPolygons();
+
+            for (const Polygon& polygon : polygons)
+            {
+                clipper.AddPath(polygon.getPoints(), join_type, ClipperLib::etClosedPolygon);
+            }
+        }
+
+        clipper.MiterLimit = miter_limit;
+
+        ClipperLib::Paths result;
+        clipper.Execute(result, static_cast<double>(distance));
+        return Shape(std::move(result));
+    }
+}
+
+void MixedLinesSet::push_back(const OpenPolyline& line)
+{
+    std::vector<PolylinePtr>::push_back(std::make_shared<OpenPolyline>(line));
+}
+
+void MixedLinesSet::push_back(OpenPolyline&& line)
+{
+    std::vector<PolylinePtr>::push_back(std::make_shared<OpenPolyline>(std::move(line)));
+}
+
+void MixedLinesSet::push_back(ClosedPolyline&& line)
+{
+    std::vector<PolylinePtr>::push_back(std::make_shared<ClosedPolyline>(std::move(line)));
+}
+
+void MixedLinesSet::push_back(const Polygon& line)
+{
+    std::vector<PolylinePtr>::push_back(std::make_shared<Polygon>(std::move(line)));
+}
+
+void MixedLinesSet::push_back(const std::shared_ptr<OpenPolyline>& line)
+{
+    std::vector<PolylinePtr>::push_back(line);
+}
+
+void MixedLinesSet::push_back(const PolylinePtr& line)
+{
+    std::vector<PolylinePtr>::push_back(line);
+}
```

Comment:
Since these all have the same operation (minus the type), I think these can be a templated method instead.

Reply:
I did that in the first place :smile: then at some point the compiler got confused because some of them are very similar, so it couldn't point to the proper one unless you made a call to it with a very explicit type, which is really ugly. So I ended up defining all the function explicitely, which is quite sad, but works. It is also a bit more explicit, this way you know what you can call.